### PR TITLE
Pulumi Apps

### PR DESCRIPTION
**Pull request by adrians5j**
## Changes
An umbrella PR that connects all of the smaller Pulumi Apps-related PRs. These PRs should be reviewed one-by-one, starting from #2476.

## TODOs
- [x] 1. https://github.com/webiny/webiny-js/pull/2476
- [x] 2. https://github.com/webiny/webiny-js/pull/2477
- [x] 3. https://github.com/webiny/webiny-js/pull/2479
- [x] 4. https://github.com/webiny/webiny-js/pull/2480
- [x] 5. https://github.com/webiny/webiny-js/pull/2481
- [x] 6. https://github.com/webiny/webiny-js/pull/2482
- [x] 7. https://github.com/webiny/webiny-js/pull/2483
- [x] 8. https://github.com/webiny/webiny-js/pull/2485

### CODE REVIEW

File: .github/workflows/nextPush.yml [hunk 1]
```diff
@@ -144,6 +144,8 @@ jobs:
       AWS_ELASTIC_SEARCH_DOMAIN_NAME: ${{ secrets.AWS_ELASTIC_SEARCH_DOMAIN_NAME }}
       ELASTIC_SEARCH_ENDPOINT: ${{ secrets.ELASTIC_SEARCH_ENDPOINT }}
       ELASTIC_SEARCH_INDEX_PREFIX: ${{ needs.init.outputs.ts }}_
+      WCP_PROJECT_ENVIRONMENT: ${{ secrets.INTEGRATION_WCP_PROJECT_ENVIRONMENT }}
```
Comment (adrians5j) [inline .github/workflows/nextPush.yml:147]: Check this.

File: .github/workflows/nextPush.yml [hunk 2]
```diff
@@ -310,7 +312,7 @@ jobs:
 
       - name: Reduce prerendering service's process-queue schedule
         working-directory: xyz/test-project
-        run: sed -i 's/5 minutes/1 minute/g' api/pulumi/dev/prerenderingService.ts
+        run: sed -i 's/5 minutes/1 minute/g' apps/api/pulumi/dev/prerenderingService.ts
```
Comment (adrians5j) [inline .github/workflows/nextPush.yml:315]: Check this.

File: .github/workflows/v5Push.yml [hunk 3]
```diff
@@ -306,7 +308,7 @@ jobs:
 
       - name: Reduce prerendering service's process-queue schedule
         working-directory: xyz/test-project
-        run: sed -i 's/5 minutes/1 minute/g' api/pulumi/dev/prerenderingService.ts
+        run: sed -i 's/5 minutes/1 minute/g' apps/api/pulumi/dev/prerenderingService.ts
```
Comment (adrians5j) [inline .github/workflows/v5Push.yml:311]: Check this.

File: api/code/apw/executeAction/package.json [hunk 4]
```diff
@@ -1,15 +0,0 @@
-{
-  "name": "api-apw-scheduler-execute-action",
```
Comment (adrians5j) [inline api/code/apw/executeAction/package.json]: Bring back this.

File: apps/api/code/graphql/jest.config.js [hunk 5]
```diff
@@ -3,7 +3,7 @@ const { log } = require("@webiny/cli/utils");
 const { getStackOutput } = require("@webiny/cli-plugin-deploy-pulumi/utils");
 
 /**
- * `getStackOutput` retrieves all values that were exported from `api/pulumi/dev/index.ts`.
+ * `getStackOutput` retrieves all values that were exported from `apps/api/pulumi/dev/index.ts`.
```
Comment (adrians5j) [inline apps/api/code/graphql/jest.config.js:6]: Check this.

File: apps/api/code/headlessCMS/jest.config.js [hunk 6]
```diff
@@ -3,7 +3,7 @@ const { log } = require("@webiny/cli/utils");
 const { getStackOutput } = require("@webiny/cli-plugin-deploy-pulumi/utils");
 
 /**
- * `getStackOutput` retrieves all values that were exported from `api/pulumi/dev/index.ts`.
+ * `getStackOutput` retrieves all values that were exported from `apps/api/pulumi/dev/index.ts`.
```
Comment (adrians5j) [inline apps/api/code/headlessCMS/jest.config.js:6]: Check this.

File: packages/project-utils/bundling/function/index.js [hunk 7]
```diff
@@ -1,4 +1,5 @@
 module.exports.createBuildFunction = require("./createBuildFunction");
 module.exports.createWatchFunction = require("./createWatchFunction");
-module.exports.buildLambdaEdge = require("./buildLambdaEdge");
-module.exports.buildCloudFrontFunction = require("./buildCloudFrontFunction");
+
+module.exports.createBuildHandler = require("./createBuildHandler");
```
Comment (adrians5j) [inline packages/project-utils/bundling/function/index.js:4]: Remove this.

File: packages/cwp-template-aws/template/common/apps/admin/code/webiny.config.ts [hunk 8]
```diff
@@ -21,7 +21,7 @@ export default {
         async watch(options) {
             invariant(options.env, NO_ENV_MESSAGE);
 
-            const output = getStackOutput({ folder: "api", env: options.env, map: API_MAP });
+            const output = getStackOutput({ folder: "apps/api", env: options.env, map: API_MAP });
```
Comment (adrians5j) [inline packages/cwp-template-aws/template/common/apps/admin/code/webiny.config.ts:24]: Check whole cwp-template.

File: packages/cli-plugin-deploy-pulumi/utils/createProjectApplicationWorkspace.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+const ncpBase = require("ncp");
+const ncp = util.promisify(ncpBase.ncp);
+const { replaceInPath } = require("replace-in-path");
+const { createProjectApplicationWorkspace } = require("@webiny/cli/utils");
+const processHooks = require("./processHooks");
+
+module.exports = async ({ projectApplication, context, inputs }) => {
+    const hookArgs = { context, inputs, projectApplication };
+
+    await runHook({ hook: "hook-before-create-workspace", context, args: hookArgs });
```
Comment (adrians5j) [inline packages/cli-plugin-deploy-pulumi/utils/createProjectApplicationWorkspace.js:12]: Delete this hook.

File: packages/cli-plugin-deploy-pulumi/commands/watch.js [hunk 10]
```diff
@@ -33,9 +44,20 @@ module.exports = async (inputs, context) => {
             cwd: path.join(process.cwd(), inputs.folder)
         });
 
-        // If exists - read default inputs from "webiny.application.js" file.
+        // If exists - read default inputs from "webiny.application.ts" file.
         inputs = merge({}, get(projectApplication, "config.cli.watch"), inputs);
 
+        // If needed, let's create a project application workspace.
+        if (projectApplication.type === "v5-workspaces") {
+            await createProjectApplicationWorkspace({
```
Comment (adrians5j) [inline packages/cli-plugin-deploy-pulumi/commands/watch.js:52]: Should this also run before build hooks? We want to have prebuilt code in workspace too.

File: packages/serverless-cms-aws/src/website/plugins/renderWebsite.ts [hunk 11]
```diff
@@ -0,0 +1,61 @@
+const LambdaClient = require("aws-sdk/clients/lambda");
+import { CliContext } from "@webiny/cli/types";
+const { getStackOutput } = require("@webiny/cli-plugin-deploy-pulumi/utils");
+
+/**
+ * On every deployment of the Website project application, this plugin ensures all pages created
+ * with the Webiny Page Builder application are re-rendered.
+ */
+export default {
```
Comment (adrians5j) [inline packages/serverless-cms-aws/src/website/plugins/renderWebsite.ts]: Revisit this.

File: packages/serverless-cms-aws/handlers/ddb/api/pageBuilder/updateSettings/webiny.config.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+const createBuildFunction = require("../../../../createBuildFunction");
+const createWatchFunction = require("../../../../createWatchFunction");
+
+module.exports = {
```
Comment (adrians5j) [inline packages/serverless-cms-aws/handlers/ddb/api/pageBuilder/updateSettings/webiny.config.js:4]: Delete this.

File: packages/serverless-cms-aws/handlers/createWatchFunction.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
+const { createBuildFunction } = require("@webiny/project-utils");
+const path = require("path");
+const findUp = require("find-up");
+
+module.exports = config => {
+    const { cwd } = config;
+    const handlersPath = findUp.sync("handlers", { cwd, type: "directory" });
+    const distPath = findUp.sync("dist", { cwd, type: "directory" });
```
Comment (adrians5j) [inline packages/serverless-cms-aws/handlers/createWatchFunction.js:8]: Finish this.

**Review by adrians5j — COMMENTED** [review]
Leaving some comments regarding the things that need a 2nd look.

**Comment by Pavel910** [comment]
The code from this PR was merged via a different PR, so this one is no longer necessary.

**Pull request closed without merging**